### PR TITLE
Fixes DNS truncate for responses longer than 512 bytes issue

### LINKED ISSUE #17: Investigate why host-resolver eats up responses longer than 512 bytes
Request for resource records with longer replies (specifically TXT records) seems to be broken. The correct behavior for the host-resolver would be to upgrade to TCP for any response longer than `512`. 

Also, direct request for `TXT` records over `TCP` doesn't seem to work.

```
 dig +tcp -t txt suse.com
```

However, the underlying stub DNS truncates everything down to 512 bytes [here](https://github.com/rancher-sandbox/rancher-desktop-host-resolver/blob/325fb3dc92af0804f94c8e2c0da3d5e6f195ba21/pkg/dns/dns.go#L252) as a workaround, we will be separating UDP/TCP handlers to do the truncation accordingly. 

This issue was discovered while testing the following issue:
https://github.com/rancher-sandbox/rancher-desktop/issues/2141

### PR DESCRIPTION
**Pull request by Nino-K**
Fixes https://github.com/rancher-sandbox/rancher-desktop-host-resolver/issues/17

### CODE REVIEW

File: pkg/vmsock/conn_windows.go [hunk 1]
```diff
@@ -22,12 +22,12 @@ import (
...
 )
 
-func Listen() (net.Listener, error) {
-	vmGUID, err := vmGUID()
-	if err != nil {
-		return nil, fmt.Errorf("Listen, could not determine VM GUID: %v", err)
-	}
-	svcPort, err := hvsock.GUIDFromString(winio.VsockServiceID(HostListenPort).String())
+func GetVMGUID() (hvsock.GUID, error) {
+	return vmGUID()
```
Comment (mook-as) [inline pkg/vmsock/conn_windows.go]: Why not just rename and export `vmGUID` directly?

Also, exported functions should get doc comments.

File: pkg/commands/start_standalone.go [hunk 2]
```diff
@@ -77,7 +77,7 @@ func randomUDPPort() (int, error) {
 	return l.LocalAddr().(*net.UDPAddr).Port, nil
 }
 
-func run() {
+func trapSignal() {
```
Comment (mook-as) [inline pkg/commands/start_standalone.go]: Consider naming this something more obvious, like `waitForExitSignal`

File: pkg/dns/dns.go [hunk 3]
```diff
@@ -67,7 +69,7 @@ func newStaticClientConfig(ips []string) (*dns.ClientConfig, error) {
 	return dns.ClientConfigFromReader(r)
 }
 
-func newHandler(IPv6 bool, hosts map[string]string, upstreamServers []string) (dns.Handler, error) {
+func newHandler(truncate, IPv6 bool, hosts map[string]string, upstreamServers []string) (dns.Handler, error) {
```
Comment (mook-as) [inline pkg/dns/dns.go:72]: There's so many things passing in here now I wonder if just passing in `ServerOptions` would be a good idea.

If you want to make it nicer, make the subset into a `HandlerOptions` that's embedded into `ServerOptions`.

File: test/cmd/dnshammer.go [hunk 4]
```diff
@@ -94,11 +102,34 @@ func do(addr string, n int, records map[string][]string) error {
 	return nil
 }
 
-func ipToString(ips []net.IP) (out []string) {
-	for _, ip := range ips {
-		out = append(out, ip.String())
+func doTXT(addr string, n int, records map[string][]string) error {
```
Comment (mook-as) [inline test/cmd/dnshammer.go]: This is almost the same as `doA`; can we merge the two (and passing in a lookup function returning `[]string, error`, plus a compare function of type `func([]string, []string) bool`)?

File: test/cmd/dnshammer.go [hunk 5]
```diff
@@ -115,6 +146,13 @@ func compare(a, b []string) bool {
 	return true
 }
 
+func ipToString(ips []net.IP) (out []string) {
```
Comment (mook-as) [inline test/cmd/dnshammer.go]: We can drop this if `lookupARecord` returns a `[]string` instead.

File: test/cmd/dnshammer.go [hunk 6]
```diff
@@ -127,3 +165,44 @@ func lookupARecord(addr, resolverProtocol, domain string) ([]net.IP, error) {
...
 	return resolver.LookupIP(ctx, "ip4", domain)
 }
+
+func lookupTXT(domain string) ([]string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), lookupTimeout)
+	defer cancel()
+	return net.DefaultResolver.LookupTXT(ctx, domain)
+}
+
+func loadRecords(p string) map[string][]string {
```
Comment (mook-as) [inline test/cmd/dnshammer.go:132]: Why do we load them from disk, instead of just generating them here (and never hitting disk)?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: pkg/dns/dns.go (hunk 3 above)
Comment (Nino-K) [reply pkg/dns/dns.go:72]: I have a cleanup PR coming up just for `dns.go` file. Let's revisit in that PR.

File: test/cmd/dnshammer.go (hunk 6 above)
Comment (Nino-K) [reply test/cmd/dnshammer.go:132]: There is a disconnect between the test runner and the `dnsHammer` which is a standalone binary. The test generates the data in memory and writes it to the file for the `dnsHammer` to consume. Also, the records needs to be shared between the test DNS server and the `dnsHammer` so generating the records in the `dnsHammer` adds more complexity.

File: test/cmd/dnshammer.go (hunk 4 above)
Comment (Nino-K) [reply test/cmd/dnshammer.go]: I considered merging these, however, the lookup functions have a different argument signature and in addition, we would need to add more lookup funcs in the close future, e.g lookupMX, lookupSRV, lookupHost, etc. They will all have a diffrent signature, if we refactor now, we would need to decouple this shortly.

File: test/cmd/dnshammer.go (hunk 6 above)
Comment (mook-as) [reply test/cmd/dnshammer.go:132]: Right, but _why_ does the test generate the test data, instead of letting `dnsHammer` do it?

Comment (mook-as) [reply test/cmd/dnshammer.go:132]: … because it needs in both the server and dnshammer.

**Review by mook-as — CHANGES_REQUESTED** [review]
Per offline conversation, Nino is going to normalize `lookupARecord` / `lookupTXT` to both use a custom resolver, which will be passed in to the function.

File: pkg/commands/start_host_windows.go [hunk 7]
```diff
@@ -19,24 +19,50 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+// StartVsockHost attempts to start two AF_VSOCK listeners, one acting
+// for TCP (stream) and the other for UDP(datagram) it then waits for the
```
Comment (mook-as) [inline pkg/commands/start_host_windows.go]:
```suggestion
// for TCP (stream) and the other for UDP (datagram); it then waits for the
```

File: test/cmd/dnshammer.go [hunk 8]
```diff
@@ -56,36 +56,40 @@ func init() {
...
+			if err := verifyResults(net.LookupTXT, contains, n, srvAddr, path); err != nil {
 				return err
 			}
 		}
 	}
 	return nil
 }
 
-func do(addr string, n int, records map[string][]string) error {
+func verifyResults(lookup func(string) ([]string, error), assert func([]string, []string) bool, n int, addr, path string) error {
```
Comment (mook-as) [inline test/cmd/dnshammer.go]: `addr` doesn't seem to be used anymore.

**Review by mook-as — CHANGES_REQUESTED** [review]

File: test/cmd/dnshammer.go [hunk 9]
```diff
@@ -48,57 +44,64 @@ var dnshammerCmd = &cobra.Command{
...
 	}
-	logrus.Infof("Successfully tested %v records, against: %v", i, addr)
+	logrus.Infof("Successfully tested %v records", i)
 	return nil
 }
 
-func ipToString(ips []net.IP) (out []string) {
-	for _, ip := range ips {
-		out = append(out, ip.String())
+func contains(expected, actual []string) bool {
```
Comment (mook-as) [inline test/cmd/dnshammer.go:99]: So this tests that, for each string in `expected`, it's a substring of something in `actual`? That is, `contains([]string{"hello", "world"}, []string{"hello world", "completely unused"})` should pass?

That's… fine, but unexpected. Might want to document that.

File: test/cmd/dnshammer.go [hunk 10]
```diff
@@ -115,15 +118,45 @@ func compare(a, b []string) bool {
...
+	for _, line := range fileData {
+		var data = []string{}
+		for j, field := range line {
+			if j == 0 {
+				domain = field
+			} else {
+				data = append(data, field)
+			}
+			records[domain] = data
+		}
```
Comment (mook-as) [inline test/cmd/dnshammer.go]: I just realized… isn't this the same as:
```suggestion
		if len(line) > 0 {
			records[line[0]] = line[1:]
		}
```

**Review by mook-as — CHANGES_REQUESTED** [review]
Just need some clarification in `contains`.

File: test/cmd/dnshammer.go (hunk 9 above)
Comment (Nino-K) [reply test/cmd/dnshammer.go:99]: Well, it's more like `contains([]string{"hello", "world"}, []string{"helloworld"})`... this is due to the way go TXT lookup resolver concatenates the strings.

**Review by mook-as — APPROVED** [review]

**Pull request merged**